### PR TITLE
fix(augmentation): accept a size-1 axis in CenterCrop3D and RandomCrop3D

### LINKED ISSUE #4705: CenterCrop3D / RandomCrop3D raise _LinAlgError for a crop with a size-1 axis
## Summary

`CenterCrop3D` and `RandomCrop3D` crash for any crop with a size-1 axis, whichever axis it is. The 2D crops accept the same request.

```python
import torch, kornia.augmentation as K
x = torch.rand(1, 1, 4, 5, 6)
for size in ((1, 3, 4), (2, 1, 4), (2, 3, 1)):
    for cls in (K.CenterCrop3D, K.RandomCrop3D):
        cls(size, p=1.0)(x)
# torch._C._LinAlgError: torch.linalg.solve: (Batch element 0): The solver failed because the input matrix is singular.
#   ... for all six combinations

K.CenterCrop3D((2, 3, 4), p=1.0)(x).shape        # torch.Size([1, 1, 2, 3, 4])  fine
K.CenterCrop((1, 3), p=1.0)(torch.rand(1, 1, 4, 5)).shape   # torch.Size([1, 1, 1, 3])  the 2D class is fine
K.RandomCrop((1, 3), p=1.0)(torch.rand(1, 1, 4, 5)).shape   # torch.Size([1, 1, 1, 3])
```

Measured on `main` @ ff204fb9c, torch 2.14.0, CPU float32.

## Cause

Both classes build the crop from `get_perspective_transform3d(src_box, dst_box)`. A size-1 axis makes the eight `src` (and `dst`) corners coplanar — the inclusive box `[z0, z0 + size - 1]` has zero extent — so the 3D DLT system is rank-deficient and `torch.linalg.solve` raises. A crop is an axis-aligned translation, so no perspective solve is needed to express it.

## Expected

A size-1 axis crops to a single slice, as in 2D (extracting a single slice from a volume is an ordinary request). If that is not going to be supported, the constructor should reject it with a `ValueError` naming the constraint rather than surfacing a linear-algebra error from inside the warp.

Found while reviewing #4634 (batch 6d of #4407); no Convention sentence there covers a size-1 axis, so it is not documented as behaviour.

Posted on behalf of @ducha-aiki by Claude (Fable 5.1).

### PR DESCRIPTION
**Pull request by RizgarOzan**
## What does this pull request do?

`CenterCrop3D` and `RandomCrop3D` raised `_LinAlgError` for any crop with a size-1 axis. They built the crop matrix with `get_perspective_transform3d` from the eight box vertices, and a size-1 axis makes those vertices coplanar, so the solve is singular. A crop only moves the source box onto the destination box, so both classes now build that translation directly (`_crop_translation3d` in `_3d/geometric/crop.py`). Changelog fragment: `changelog.d/4732.fixed.md`.

For crops that worked before, the matrix is the same: over 300 random `RandomCrop3D` draws the largest difference from the old `get_perspective_transform3d` result was 1.3e-14.

## Related issue or discussion

Fixes #4705

## What did you check?

```text
$ python -m pytest tests/augmentation/test_conventions_3d.py -k 4705     # main: 4 failed (_LinAlgError); this PR: 4 passed
$ python -m pytest tests/augmentation tests/geometry/transform
5 failed, 4543 passed, 3418 skipped, 33 xfailed, 5 xpassed
$ python -m pytest --doctest-modules kornia/augmentation/_3d/geometric/crop.py kornia/augmentation/_3d/geometric/center_crop.py   # 2 passed
$ python -m pytest tests/test_api_surface.py                                # 148 passed
$ ty check kornia                                                           # All checks passed!
$ pre-commit run --files <changed files>                                    # all hooks pass
```

The new test crops (1, 3, 4), (2, 1, 4), (2, 3, 1) and (1, 1, 1) from a 4x5x6 volume with both classes and compares with the slice at the sampled offset. The 5 failures are the same ones I get on `main` (the four `*_pickle_4469` tests and `test_convention_a_mask_key_is_required_only_to_derive_the_parameters`, Windows, Python 3.14, torch 2.14.0 CPU).

Not changed here:
- A size-1 output makes torch print its `affine_grid` warning about unit-size grids with `align_corners=True`. The values are right; I left the warp alone.
- The functional `center_crop3d` still goes through `get_perspective_transform3d` and raises the same error. It lives in `geometry/transform/crop3d.py`, which #4556 is changing, so I kept this PR to the two classes.
- While checking the 2D side: `CenterCrop((1, 3), cropping_mode="resample")` returns NaN on a 4x5 image. The default `slice` mode is fine. I can open an issue for that if it is useful.

## How did AI help?

I made this with Claude Code (an AI coding agent). It reproduced the issue, wrote the change and the test, and ran every check listed above. The numbers in this description come from those runs.

### CODE REVIEW

**Comment by RizgarOzan** [comment]
The two half-precision jobs failed on the new test, my miss: I only ran it in float32. On CPU, `F.affine_grid` builds a size-1 axis with `torch.tensor`, which has no float16/bfloat16 kernel (`"tensor_cpu" not implemented for 'Half'`), so the crop itself works but the warp can't run there. 11c6de5d adds a `supports_unit_size_3d_affine_grid` probe next to the other kernel probes in `testing/base.py` and skips on it. Locally `tests/augmentation/test_conventions_3d.py` passes in float32/float64 and skips those 4 cases in float16/bfloat16; `tests/augmentation tests/geometry/transform` with `--known-failure-profile=cpu-float16 --xfail-known-failures` gives the same 5 failures as main (the Py 3.14 pickle ones and the Transplantation message).

**Review by ducha-aiki — APPROVED** [review]
## Review at `11c6de5db8039604830aa5d844c7415eefd2a766`

Approve. I reviewed the branch and its merge with `origin/main` @ 8b03a8085. The branch is based on #4691's merge commit, so the merge brings in only #4665, #4681, #4721 and #4728, and none of them touches these files.

**What I verified**

- **Same matrix as before on every crop that already worked.** Over 200 random volumes and sizes (every axis ≥ 2), for both `RandomCrop3D` and `CenterCrop3D` in float32 and float64, `_crop_translation3d` differs from the old `get_perspective_transform3d(src, dst)` by at most 3.2e-14. The output matches the slice at the sampled offset to within 9e-7, which is the existing warp noise.
- **The new pin fails on main.** With the PR's test file and `testing/base.py` copied onto main, `-k 4705` gives 4 failed (`_LinAlgError`). On the merge tree it gives 8 passed. The other 8 are float16/bfloat16 skips.
- **The half skip is justified.** On CPU torch 2.14, `CenterCrop3D((1, 3, 4))` in float16 or bfloat16 raises `"tensor_cpu" not implemented for 'Half'` from `affine_grid`, while `(2, 3, 4)` works. That is a torch kernel gap. The old path failed on the same input anyway.
- **Mutation testing.** I ran five one-token mutants of `_crop_translation3d` against the 3D augmentation tests. All five are killed: sign flip (22 failures), the wrong vertex `src[:, 6]` (27), off by +1 (28), reversed axis order (18), and no translation (22).
- **Interaction with #4691's one-voxel rejection.** I checked this on a depth-1 input (`1×5×6`). Size-1 crops are accepted and take the right slab: `(1, 3, 4)` and `(1, 1, 1)` in both classes, and `(1, 5, 6)`, which covers the whole volume. One voxel too large still raises: `(2, 3, 4)` and `(1, 6, 6)` get the `ValueError` from `RandomCrop3D` and the `AssertionError` from `CenterCrop3D`. Padding plus #4691's per-channel `fill` with a size-1 depth works. A batch of 4 with per-sample offsets matches each sample's slice.
- **No regressions.** `tests/augmentation tests/geometry/transform` gives 4554 passed on the merge tree and 4550 on main. The skips (3420), xfails (32) and xpasses (5) are the same on both, and there are no failures. CI is 41/41 green.
- **Housekeeping.** The fragment is a correct `fixed`. `Fixes #4705` is accurate, since the issue names exactly these two classes. The only `#4705` hit in `kornia/` is the helper's rationale line.

**Copilot threads.** I verified both and dropped them. A `_`-prefixed helper shared inside `_3d/geometric` is normal in kornia. The docstring already says the matrix only moves the first vertex, which is the pure-translation assumption.

**Filed.** Your notes on the 2D `CenterCrop((1, 3), cropping_mode="resample")` NaN and the functional `center_crop3d` are now #4747. I confirmed both on main, and the issue also covers `crop_by_boxes`, `crop_and_resize` and `crop_and_resize3d`, which have the same root cause. Nothing there blocks this PR.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

File: kornia/augmentation/_3d/geometric/center_crop.py [hunk 1]
```diff
@@ -21,8 +21,9 @@
 
 from kornia.augmentation import random_generator as rg
 from kornia.augmentation._3d.geometric.base import GeometricAugmentationBase3D
+from kornia.augmentation._3d.geometric.crop import _crop_translation3d
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/augmentation/_3d/geometric/center_crop.py:24]: Dropped: the helper is shared between the two 3D crop modules of the same private package (`kornia.augmentation._3d`) and is not exported. Checked in the approving review.

Posted on behalf of @ducha-aiki by Claude (Opus 5).

File: kornia/augmentation/_3d/geometric/crop.py [hunk 2]
```diff
@@ -25,7 +25,18 @@
...
+
+
+def _crop_translation3d(src: torch.Tensor, dst: torch.Tensor) -> torch.Tensor:
+    """Return the matrix of a crop, which only moves the first ``src`` vertex onto the first ``dst`` vertex.
+
+    Solving the perspective system from the eight vertices instead fails for a crop with a size-1 axis, whose
+    vertices are coplanar (#4705).
+    """
+    transform = torch.eye(4, device=src.device, dtype=src.dtype).repeat(src.shape[0], 1, 1)
+    transform[:, :3, 3] = dst[:, 0] - src[:, 0]
```
[the remark this answers is not in the document]
Comment (ducha-aiki) [reply kornia/augmentation/_3d/geometric/crop.py:38]: Dropped: five mutants of the helper (sign flip, wrong vertex, off-by-one, reversed axes, no translation) each fail 18–28 tests, and the translation matrix matches the previous solve to 3.2e-14 on every crop that already worked. Checked in the approving review.

Posted on behalf of @ducha-aiki by Claude (Opus 5).
